### DISCUSSION TITLE
Full cluster shutdown/reboot

### DISCUSSION THREAD

**Discussion by bo0tzz** [cloudnative-pg/cloudnative-pg#166] [General]
Hi there,

I'm using Kubernetes in a homelab environment and would really like to use this operator for my database deployment. One unusual factor there is that occasionally the entire Kubernetes cluster will be (cleanly) shut down, for example if I'm moving house. I'm trying to ensure that I can do that safely, without risking data loss. Am I correct in thinking that the following procedure would work?

0. Have adequate backups
1. Set a `nodeMaintenanceWindow` with `reusePVC: true`
2. Cordon and drain all worker nodes
3. Shut down all worker nodes
4. Shut down the master node 
 
Then for starting back up, I would follow these steps in reverse. Is there anything important I'm missing here, or would that be an adequate approach? Thanks!

**Comment by sxd** [comment]
Hi @bo0tzz 

That's something that we haven't think about, yet, but if you can give us a hand testing, will be awesome.
Related to the steps to follow, looks precise, probably you can actually start using MinIO for continuous backup, meaning that you don't need to create a backup, doing this the operator should handle the recover without issues, but again, we haven't test this use case so feedback is required!

In the meantime I'm converting this into a discussion, since it's the new GitHub way to handle this kind of discussions and use cases, then, we can open issues/bugs starting from this discussion.

Regards!
.

**Comment by bo0tzz** [comment]
Very nice! After I've set things up I'll kick it around a bit and report back with how it holds up 👍

**Reply by gbartolini** [reply]
You can also fence the entire cluster (look for fencing in the docs - I’m on the phone now)

**Comment by bo0tzz** [comment]
I set up a lab cluster and gave it a poke. The scope shifted slightly as I was messing around, but I figured I'd share my notes here anyways just in case they end up being useful. 

I set up k3s on ubuntu 22.04 with 1 leader and 3 worker nodes. I deployed minio gateway backed by an nfs volume on another VM for the WAL archive, [local path provisioner](https://github.com/rancher/local-path-provisioner) for the cluster pods' volumes, and the CNPG operator with the default settings. I used a 3 instance PG cluster for everything. I also ran a script to insert some dummy data every .2 seconds, just to get a feel for how things behaved.

My first test was to run `shutdown -h now` simultaneously on all the Kubernetes nodes, without preparing the database in any way. Things seemed to shut down gracefully, everything came back up with the data intact, and nothing complained at me 👍 .

Next I decided to get a bit rougher - I deleted all the cluster pods and their attached volumes. I expected CNPG to catch this and rebuild from the WAL archive, but instead it just sat there and did nothing. `kubectl cnpg status` reported:
```
Status:            Cluster in healthy state
Instances:         3
Ready instances:   0
(...)
Instances status
[empty]
```

I couldn't get the operator to pick up on the fact that the whole cluster was gone, so I deleted the Cluster object and tried to bootstrap a new cluster from the WAL archive using `externalCluster`  (my notes don't say anything about recreating the old Cluster object - I don't remember if I tried that).
When trying to bootstrap the new cluster, I got a `no target backup found` error. I had expected that I could just restore from the WAL archive, but it seems that assumption was wrong?

So I wiped things clean, created a new cluster, and created a Backup for this one. Now, deleting the cluster and bootstrapping from that backup worked as expected, including the restore of data that was inserted after the backup from the WAL log. 

At this point I was confident enough that I didn't actually try the `nodeMaintenanceWindow` or fencing for a cluster shutdown - I can imagine they'd be useful for single-node shutdowns, but it seems like a whole-cluster shutdown is fine without them. 

My main takeways are:
1. I'm happy shutting things down without much prep.
2. The CNPG operator doesn't notice things as much as I'd hope (but has been very nice to use otherwise).
3. It'd be nice if clusters could bootstrap from just a WAL archive or, if that isn't possible, if the docs had a note that you need a Backup as well.

I hope this is useful for you!
